### PR TITLE
Fence the phantom LEFT JOIN row everywhere it can be seen (EXP-25/26/28/29, PBT-13)

### PR DESCRIPTION
**Pull request by anentropic**
Stacked on #215. The largest correctness win of the round — twelve wrong numbers, each reproduced individually before fixing.

## The class

Every synthesized join is a `LEFT JOIN` from the base table, so a base row with no match (an order with no line items) survives as a row of NULLs. #203 fenced that row for `COUNT(*)` and for a whitelist of five aggregates over a *recognized constant literal*. **The whitelist leaked four ways**, and the entry's own rationale was wrong.

| shape | returned | correct |
|---|---|---|
| `COUNT(DISTINCT 1)` | 1 | **0** |
| `COUNT(1+0)` | 1 | **0** |
| `MIN(1)` | 1 | **NULL** |
| `SUM(COALESCE(li.qty, 99))` | 99 | **NULL** |
| `SUM(<parent fact>)` at child grain | 25 | **20** |

TECH-DEBT #56 excused `COUNT(DISTINCT <const>)` and `MIN`/`MAX`/`AVG` as "multiplicity-invariant". True, and irrelevant: the phantom row is not a duplicate, it is **a row that should not exist**. That entry is corrected here.

## The fix — a guard, not a bigger whitelist

`guard_aggregate_args` drops per-argument classification entirely. On a non-base source table with a declared PK, the **first argument of every aggregate** is wrapped as `AGG([DISTINCT ]CASE WHEN <pk> IS NOT NULL THEN <arg> END)`. The wrap is neutral on real rows *whatever the argument is* — the PK is non-NULL there — which is exactly why it no longer needs to recognize the argument.

Verified empirically against DuckDB 1.5.5 that every multi-argument aggregate drops a row whose first argument is NULL, so guarding arg 1 suffices and `STRING_AGG`'s separator stays a constant of the call. In-call `ORDER BY` and outer `FILTER` are copied verbatim; `COUNT(*)` is skipped so SG-8's star rewrite still owns it. Function names now match as whole identifier words, so `sum_no_overflow(` is reachable at all (it never was).

## Row-level shapes (EXP-28/29)

FACTS queries and dims-only `DISTINCT` returned a manufactured all-NULL row per childless parent — indistinguishable from a genuine NULL. Fixed with `WHERE <pk> IS NOT NULL` when every queried member lives on one table **strictly below the base**.

**That direction condition is the most important detail here, and it was not in the first cut.** Filtering for any single non-base table made `multi_hop_join_proptest` fail immediately: a member *above* the base is an attribute of each base row, so a base row with a NULL or dangling foreign key is a genuine row of the view and its NULL attribute is part of the answer. Both directions are now pinned by unit tests.

## Coverage

28 tests in the new `src/expand/tests_phantom_row_guard.rs` (12 numeric reds observed individually), plus `test/sql/cr20260808_phantom_row_guard.test`.

**PBT-13** — `child_grain_proptest` re-introduced `where_clause: None`, the canonical blind-spot shape, two commits after CLAUDE.md enshrined it. Now un-pinned: predicates over both sides of the join (including one that empties a real parent's child set, making it childless *at filter time*), all four escaped aggregate shapes, a parent-fact reference metric, and FACTS requests with a row-level oracle. Five anti-vacuity guards. Verified red with the fix reverted.

Residuals (NULL-retaining aggregates, no-PK tables, multi-table row-level queries) are recorded in new TECH-DEBT #67 rather than left implicit.

---
_Generated by [Claude Code](https://claude.ai/code/session_016PT3CfUdQoRfc2SnvdzrbB)_

### CODE REVIEW

**Comment by anentropic** [comment]
Thanks — one real catch fixed, and I verified the MSRV concern rather than assuming.

**Stale ledger references — fixed.** Both `facts.rs:691` and `sql_gen.rs:218` pointed at TECH-DEBT #61. Good catch: the entries were renumbered late (the still-open findings carried across previous rounds took #61–#66), and the code comments kept the draft number, so they were sending readers to the *catalog schema-identity* entry instead of the phantom-row residuals. Both now say #67, and I grepped the tree for other survivors — none.

**`Option::is_none_or` — checked, not an issue.** There's no `rust-version` key in `Cargo.toml`, so the project declares no MSRV, and `rust-toolchain.toml` pins `1.97.1`. `is_none_or` stabilised in 1.82, well below the pin. Worth flagging in general, but not a live risk here — and if an MSRV policy is ever adopted, the pin is the thing that would need to move first.

**Pointer arithmetic for `inner_start` — agreed in principle, deliberately not now.** Having `scan_aggregate_calls` hand explicit byte indices to the visitor is the better interface, and I'd take it. But this specific code is the guard that just replaced a leaking whitelist and closed five numerically-confirmed wrong numbers; refactoring its index plumbing in the same change trades a verified implementation for an unverified one, and it sits mid-stack where every edit re-rolls the six PRs above it. The current form is covered by 28 tests including the `ORDER BY` / `FILTER` / `DISTINCT` splice edge cases that are exactly what would break. Better as a follow-up with its own before/after.

CI red here was a real defect I'd missed: the new tests execute SQL against in-memory DuckDB, which the `extension` feature's loadable stubs can't provide, so the whole module is now gated the way `semi_additive.rs`'s `execution` submodule is. Verified both configurations — extension-feature run green, and all 28 tests still execute under default features.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/expand/sql_gen.rs [hunk 1]
```diff
@@ -183,6 +183,113 @@ fn resolve_names<'a, T: Resolvable, N: AsRef<str>>(
...
+/// and its NULL attribute is part of the answer, not a join artifact. Filtering
+/// there would delete real rows — `multi_hop_join_proptest` catches it
+/// immediately. Below the base the situation is reversed: the base row has no
+/// counterpart at the finer grain at all, so the NULL-extended row is
+/// manufactured.
+///
+/// When the members span several tables the natural grain is ambiguous (a base
+/// member legitimately keeps its row even with no child), and when the table
+/// declares no PRIMARY KEY there is no column that distinguishes a phantom;
+/// both keep today's behaviour. See TECH-DEBT #61.
```
[the remark this answers is not in the document]
Comment (anentropic) [reply src/expand/sql_gen.rs]: Right — the comments were written against #61 before #67 existed. Fixed in commit `694dc9e`, which repoints the phantom-row residual comments (here and in `facts.rs`) at TECH-DEBT #67.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/expand/facts.rs [hunk 2]
```diff
@@ -474,102 +563,225 @@ fn is_constant_literal(arg: &str) -> bool {
...
-/// `pk_ref`, so a NULL-extended LEFT JOIN row contributes nothing (EXP-21).
+/// Byte offset just past a leading `DISTINCT` / `ALL` set quantifier in an
+/// aggregate's argument list, or 0 when there is none.
+fn leading_quantifier_end(inner: &str) -> usize {
+    let lead = inner.len() - inner.trim_start().len();
+    let rest = &inner.as_bytes()[lead..];
+    for kw in [b"DISTINCT".as_slice(), b"ALL".as_slice()] {
+        if rest.len() >= kw.len()
+            && rest[..kw.len()].eq_ignore_ascii_case(kw)
+            && rest.get(kw.len()).is_none_or(|b| is_word_boundary_char(*b))
```
[the remark this answers is not in the document]
Comment (anentropic) [reply src/expand/facts.rs:621]: No MSRV risk here — this project doesn't carry an MSRV to break.

`rust-toolchain.toml` pins an exact toolchain (`channel = "1.97.1"`, bumped by Dependabot), and `Cargo.toml` declares no `rust-version`, so there is no older-compiler contract to honour. `Option::is_none_or` is also already load-bearing in five places that predate this PR — `graph/derived_metrics.rs`, `expand/semi_additive.rs`, `body_parser/mod.rs`, and `body_parser/scan.rs` — so adopting the `match`/`map_or` spelling here would make this call site inconsistent with the surrounding code rather than safer.

Keeping `is_none_or`: it reads better than `map_or(true, …)` at a boundary check, and CI compiles the crate on the pinned toolchain on every PR, so a genuine availability problem would be a hard build failure rather than something that slips through.

(Same for the `child_grain_proptest.rs` comment.)

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tests/child_grain_proptest.rs [hunk 3]
```diff
@@ -377,3 +723,89 @@ fn generator_selects_constant_argument_metrics() {
...
+            let children: Vec<Option<i64>> = case
+                .inst
+                .li_rows
+                .iter()
+                .filter(|(order_id, _)| *order_id == Some(id))
+                .map(|(_, amount)| *amount)
+                .collect();
+            // A real parent (has children) all of whose children fail the
+            // simplest child-side test the generator can emit.
+            !children.is_empty() && children.iter().all(|a| a.is_none_or(|v| v <= 0))
```
[the remark this answers is not in the document]
Comment (anentropic) [reply tests/child_grain_proptest.rs:804]: Same as the `facts.rs` thread: no MSRV to break here. `rust-toolchain.toml` pins `1.97.1` exactly, `Cargo.toml` declares no `rust-version`, and `is_none_or` already appears in several modules that predate this PR. Full reasoning there.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/expand/facts.rs [hunk 4]
```diff
@@ -474,102 +563,225 @@ fn is_constant_literal(arg: &str) -> bool {
...
+        }
+        out.push_str("CASE WHEN ");
+        out.push_str(pk_ref);
+        out.push_str(" IS NOT NULL THEN ");
+        out.push_str(inner[quantifier_end..arg_end].trim());
+        out.push_str(" END");
+        copied = inner_start + arg_end;
+        changed = true;
+        false // never stop early: every call has to be guarded
+    });
```
[the remark this answers is not in the document]
Comment (anentropic) [reply src/expand/facts.rs:718]: Fair observation, and I agree the index-passing signature is the better shape. Not taking it in this PR, but the reasoning is narrower than "won't fix":

The arithmetic is sound as written — `scan_aggregate_calls` cuts `inner` from `expr` itself, so `inner.as_ptr() - expr.as_ptr()` is a real byte offset into `expr`, and the comment above it says exactly that. The fragility you're pointing at is real but *latent*: it only bites if someone later changes `scan_aggregate_calls` to hand the visitor a slice that isn't a subslice of `expr` (an owned `String`, or one built from a normalized copy). At that point the subtraction becomes meaningless — and, worse, wouldn't panic, it would silently compute a nonsense offset and splice at the wrong place.

Two reasons to leave it here rather than fix it now: this PR is mid-way through a stacked merge train with green CI, and changing the visitor signature would touch the other caller (`has_constant_arg_aggregate`) that exists specifically so the two paths cannot disagree about what counts as an aggregate call — so it wants its own change with its own test run, not a drive-by at merge time.

Not opening a TECH-DEBT entry because this isn't a knowingly-degraded behaviour — the emitted SQL is correct today, and the project's rule is about shipped behaviour, not internal shape. Flagging it here so it's on record if that function is revisited.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/expand/facts.rs [hunk 5]
```diff
@@ -474,102 +563,225 @@ fn is_constant_literal(arg: &str) -> bool {
...
+/// Returns `None` when the expression contains no guardable aggregate call,
+/// matching [`rewrite_count_star`]'s contract. Quoted regions are inert
+/// throughout ([`crate::util::QuoteState`]), so `'count(1)'` and
+/// `"sum(1) col"` are left alone.
+///
+/// Two residuals are recorded in TECH-DEBT #61: the NULL-RETAINING aggregates
+/// (`list` / `array_agg` / `first` / `last` / `arbitrary`, which keep a NULL
+/// element rather than skipping it, so the phantom leaves a NULL in the list
+/// instead of a value), and the case where the source table declares no
+/// PRIMARY KEY, where there is no column to guard with at all.
```
[the remark this answers is not in the document]
Comment (anentropic) [reply src/expand/facts.rs]: Agreed — fixed in commit `694dc9e`, which repoints these residual comments at TECH-DEBT #67.

---
_Generated by [Claude Code](https://claude.ai/code)_
